src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -11,32 +11,62 @@ namespace System.Net
 {
     internal class IPAddressParser
     {
+
+        internal static unsafe bool ContainsColon(char* ipStringPtr, int ipStringPtrLength)
+        {
+            for (var i = 0; i < ipStringPtrLength; i++)
+            {
+                if (ipStringPtr[i] == ':')
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
         internal static unsafe IPAddress Parse(string ipString, bool tryParse)
         {
-            if (ipString == null)
+            fixed (char* ipStringPtr = ipString)
+            {
+                return Parse(ipStringPtr, ipString.Length, tryParse);
+            }
+        }
+
+        internal static unsafe IPAddress Parse(ReadOnlySpan<char> ipSpan, bool tryParse)
+        {
+            fixed (char* ipStringPtr = &ipSpan.DangerousGetPinnableReference())
+            {
+                return Parse(ipStringPtr, ipSpan.Length, tryParse);
+            }
+
+        }
+   
+        internal static unsafe IPAddress Parse(char* ipStringPtr, int ipStringPtrLength, bool tryParse)
+        {
+            if (ipStringPtr == null)
             {
                 if (tryParse)
                 {
                     return null;
                 }
-                throw new ArgumentNullException(nameof(ipString));
+                throw new ArgumentNullException(nameof(ipStringPtr));
             }
 
-            if (ipString.IndexOf(':') != -1)
+            if (ContainsColon(ipStringPtr, ipStringPtrLength))
```

Comment:
If you do switch to using span instead of pointers, you shouldn't need this custom ContainsColon either, as you should be able to just use Span.IndexOf.  At the moment that'll require a reference to System.Memory, but hopefully that won't be needed in the near future.